rfcs/ui-building-blocks.md

```diff
@@ -0,0 +1,472 @@
+# Feature Name: `ui_building_blocks`
+
+## Summary
+
+This RFC establishes a common vocabulary and basic data structures for UI.
+It is combined with a design pattern to manipulate UI styles ergonomically using Bevy's ECS, to tangibly demonstrate the feasibility of the approach.
+
+## Motivation
+
+UI is the next big challenge for Bevy, with a complex set of data structures that need to be stored and manipulated.
+In particular, the ability to style elements in a reusable fashion is central to a polished, easy-to UI, but there's no immediately obvious approach to handling this.
+What makes a "style" in Bevy? How do we represent the individual attributes? How are widgets represented? How are styles composed? How are they applied?
+
+None of these questions are terribly challenging to implement on their own, but taking the first-seen path is particularly dangerous here, due to the risk of proliferation of non-local configuration and similar-but-not-identical attributes, as seen in CSS.
+
+We need to settle on a common standard, in order to enable interop between UI crates and establish the basic building blocks that more complex UI decisions can build off of.
+
+## Guide-level explanation
+
+At heart, user interfaces (UI) in Bevy are "just another part of the game".
+It's built out of the same core primitives that you'll see when building your game logic: entities, components, systems, scenes, events and so on.
+That however doesn't mean it's a complete free-for-all: `bevy_ui` defines several built-in components, systems and design patterns
+to make working with UI ergonomic, maintainable and interoperable without making it any harder to extend or hook into.
+
+Commonly, UI is defined by three core components, each controlling a certain type of behavior:
+
+1. `ScreenSpace`, which dictates that an entity's transform should be drawn relative to the player's screen, rather than the world.
+Entities with the `ScreenSpace` component are drawn by the UI camera; all others are drawn with ordinary cameras.
+2. `Layout`, which tells Bevy's built-in layout systems to control its positioning relative to other `Layout` entities.
+The values of this component's fields determine exactly how this is done.
+3. `Widget`, a general purpose marker component that designates an entity as "part of the UI".
```

Comment:
Would `UIComponent` be more appropriate for this naming convention? Or just `UI`? Or... `UserInterface`?
Widget seems here to make this domain specific just to Bevy. Is that desired?

An instance of when disambiguation might be needed with the term "widget":
[iOS widgets](https://support.apple.com/en-us/HT207122) and [Android widgets](https://developer.android.com/guide/topics/appwidgets/overview) already exist. Would there ever be a future for Bevy where both of these "widget" terms come into conflict? Or need disambiguation?